common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/NfcSmartcardCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,170 @@
+                            @Override
+                            public void onGetSession(@NonNull final ISmartcardSession session) throws Exception {
+                                tryUsingSmartcardWithPin(pin, certDetails, request, session);
+                                clearPin(pin);
+                            }
+
+                            @Override
+                            public void onException(@NonNull final Exception e) {
+                                indicateGeneralException(methodTag, e);
+                                request.cancel();
+                                clearPin(pin);
+                                mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+                            }
+                        });
+                    }
+
+                    @Override
+                    public void onClosedConnection() {
+                        //Nothing needed.
+                    }
+                });
+                mSmartcardCertBasedAuthManager.startDiscovery(mActivity);
+            }
+        };
+    }
+
+    /**
+     * Checks to see if PIN for smartcard is correct.
+     * If so, proceed to attempt authentication.
+     * Otherwise, handle the incorrect PIN based on how many PIN attempts are remaining.
+     * @param pin char array containing PIN attempt.
+     * @param certDetails ICertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @param session An ISmartcardSession created to help with interactions pertaining to certificates.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    protected void tryUsingSmartcardWithPin(@NonNull final char[] pin,
+                                            @NonNull final ICertDetails certDetails,
+                                            @NonNull final ClientCertRequest request,
+                                            @NonNull final ISmartcardSession session)
+            throws Exception {
+        final String methodTag = TAG + ":tryUsingSmartcardWithPin";
+        if (session.verifyPin(pin)) {
+            //If pin is successfully verified, we will use the certificate to perform the rest of the logic for authentication.
+            useSmartcardCertForAuth(certDetails, pin, session, request);
+            return;
+        }
+        final int attemptsRemaining = session.getPinAttemptsRemaining();
+        mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+        // If the number of attempts is 0, no more attempts will be allowed.
+        if (attemptsRemaining == 0) {
+            //We must display a dialog informing the user that they have made too many incorrect attempts,
+            // and the user will need to figure out a way to reset their key outside of our library.
```

Comment:
nit: I think the comment 
// If the number of attempts is 0, no more attempts will be allowed.
and
//We must display a dialog informing the user that they have made too many incorrect attempts,
// and the user will need to figure out a way to reset their key outside of our library.
can be removed, as the code is clear enough().

(maybe rename indicateTooManyFailedAttempts to promptTooManyFailedPinAttempts()?